src/relax/analysis/dataflow_analysis.cc

```diff
@@ -0,0 +1,249 @@
+    preds->push_back({final_true_idx, final_false_idx});
+    succs->push_back({merge_idx + 1});
+    // update the successors of the final true and false indices as well
+    succs->at(final_true_idx).push_back(merge_idx);
+    succs->at(final_false_idx).push_back(merge_idx);
+  }
+  // move on to next binding
+  size_t next_block_idx = block_idx;
+  size_t next_binding_idx = binding_idx + 1;
+  if (next_binding_idx >= seq->blocks[block_idx]->bindings.size()) {
+    next_block_idx = block_idx + 1;
+    next_binding_idx = 0U;
+  }
+  return ExtractCFGHelper(seq, {}, next_block_idx, next_binding_idx, {bindings->size() - 1},
+                          bindings, preds, succs);
+}
+
+ControlFlowGraph ExtractCFG(const Function& func) {
+  std::vector<GraphBinding> bindings;
+  std::vector<std::vector<size_t>> preds;
+  std::vector<std::vector<size_t>> succs;
+  ExtractCFGHelper(Downcast<SeqExpr>(func->body), func->params, 0U, 0U, {}, &bindings, &preds,
+                   &succs);
+
+  Array<Array<Integer>> pred_arr;
+  for (auto pred_vec : preds) {
+    Array<Integer> pred_ints;
+    for (auto idx : pred_vec) {
+      pred_ints.push_back(Integer(idx));
+    }
+    pred_arr.push_back(pred_ints);
+  }
+  Array<Array<Integer>> succ_arr;
+  for (auto succ_vec : succs) {
+    Array<Integer> succ_ints;
+    for (auto idx : succ_vec) {
+      succ_ints.push_back(Integer(idx));
+    }
+    succ_arr.push_back(succ_ints);
+  }
+  return ControlFlowGraph::Create(Array<GraphBinding>(bindings), pred_arr, succ_arr);
+}
+
+std::pair<Array<ObjectRef>, Array<ObjectRef>> DataflowAnalysis(
+    const ControlFlowGraph& cfg, const ObjectRef& init,
+    std::function<ObjectRef(const GraphBinding&, const ObjectRef&)> transfer_func,
+    std::function<ObjectRef(const ObjectRef&, const ObjectRef&)> merge_func, bool forward) {
+  std::vector<ObjectRef> in_map;
+  std::vector<ObjectRef> out_map;
+  for (size_t i = 0; i < cfg->bindings.size(); i++) {
+    in_map.push_back(init);
+    out_map.push_back(init);
+  }
+
+  // Modification from Adrian Sampson's version:
+  // Since there are no loops in our AST, one traversal through the CFG suffices.
+  // We will do BFS
+  std::queue<size_t> worklist;
+  worklist.push((forward) ? 0 : cfg->bindings.size() - 1);
+  while (!worklist.empty()) {
+    size_t idx = worklist.front();
+    worklist.pop();
+    Array<Integer> prev = (forward) ? cfg->preds[idx] : cfg->succs[idx];
+    Array<Integer> next = (forward) ? cfg->succs[idx] : cfg->preds[idx];
+    std::vector<ObjectRef>* results = (forward) ? &out_map : &in_map;
+    std::vector<ObjectRef>* inputs = (forward) ? &in_map : &out_map;
+
+    // Cases (for forward analysis):
+    // 0 predecessors: The first block in the function
+    // 1 predecessor: A branch in an If node (no merge needed)
+    // 2 predecessors: The merge block after an If node (merge needed)
+    // (Analogous for successors in backward analysis)
+    inputs->operator[](idx) = (prev.size() == 0)   ? init
```

Comment:
Nit: If this is declared as `std::vector<ObjectRef>& inputs = (forward)? out_map : in_map;`, then the LHS of the assignment becomes `inputs[idx]` instead of `inputs->operator[](idx)`.